services/venice-controller/src/main/java/com/linkedin/venice/controller/kafka/protocol/serializer/SemanticDetector.java

```diff
@@ -0,0 +1,438 @@
+   * @param defaultValue The default value of the object
+   * @throws VeniceProtocolException if there is a semantic change between the two schemas & value from @code{@param object} is non-default value
+   * OR if the object does not match any schema within the union schemas.
+   */
+  public static void traverseUnion(
+      Object object,
+      Schema currentSchema,
+      Schema targetSchema,
+      String name,
+      Object defaultValue) {
+    List<Schema> schemasInUnion = currentSchema.getTypes();
+    Schema objectSchema = getObjectSchema(object, currentSchema);
+
+    if (objectSchema == null) {
+      throw new VeniceProtocolException(
+          String.format(
+              "Field %s: Object %s does not match any schema within union schemas %s",
+              formatFieldName(name),
+              object,
+              schemasInUnion));
+    }
+
+    Map<String, Schema> targetSchemaMap =
+        targetSchema.getTypes().stream().collect(Collectors.toMap(Schema::getName, s -> s));
+
+    Schema subTargetSchema = targetSchemaMap.get(objectSchema.getName());
+    String objectName = buildFieldPath(name, objectSchema.getName());
+    traverseAndValidate(object, objectSchema, subTargetSchema, objectName, defaultValue);
+  }
+
+  public static Schema getObjectSchema(Object object, Schema unionSchema) {
+    for (Schema schema: unionSchema.getTypes()) {
+      if (isCorrectSchema(object, schema)) {
+        return schema;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Check if the value matches correct schema type.
+   * @param value the value to check
+   * @param schema the schema to check against
+   * @return true if the value matches the schema type, false otherwise
+   */
+  private static boolean isCorrectSchema(Object value, Schema schema) {
+    switch (schema.getType()) {
+      case STRING:
+        return value instanceof String;
+      case INT:
+        return value instanceof Integer;
+      case LONG:
+        return value instanceof Long;
+      case FLOAT:
+        return value instanceof Float;
+      case DOUBLE:
+        return value instanceof Double;
+      case BOOLEAN:
+        return value instanceof Boolean;
+      case ENUM:
+        return value instanceof String && schema.getEnumSymbols().contains(value);
+      case FIXED:
+        return (value instanceof byte[]) && ((byte[]) value).length == schema.getFixedSize();
+      case BYTES:
+        return value instanceof byte[];
+      case RECORD:
+        return value instanceof GenericRecord && ((GenericRecord) value).getSchema().equals(schema);
+      case ARRAY:
+        return value instanceof List
+            && ((List<?>) value).stream().allMatch(element -> isCorrectSchema(element, schema.getElementType()));
+      case MAP:
+        return value instanceof Map && ((Map<?, ?>) value).values()
+            .stream()
+            .allMatch(element -> isCorrectSchema(element, schema.getValueType()));
```

Comment:
We probably don't need to loop through the entire collection in this check. `anyMatch` should be good enough. WDYT?

Reply:
`anyMatch` works. Updating.